### ISSUE TITLE
[BUG] Restarting postfix fails

### ISSUE THREAD

**Issue by casperklein** [docker-mailserver/docker-mailserver#3116]
### Miscellaneous first checks

- [X] I checked that all ports are open and not blocked by my ISP / hosting provider.
- [X] I know that SSL errors are likely the result of a wrong setup on the user side and not caused by DMS itself. I'm confident my setup is correct.

### Affected Component(s)

`supervisorctl restart postfix` fails

### What happened and when does this occur?

Running `supervisorctl restart postfix` inside the container, leads to the following endless looping error messages in `mail.log`:

```
2023-02-25 22:51:45,196 INFO waiting for postfix to stop
2023-02-25 22:51:45,197 INFO stopped: postfix (terminated by SIGTERM)
2023-02-25 22:51:45,199 INFO spawned: 'postfix' with pid 4682
2023-02-25 22:51:45,200 INFO success: postfix entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
Feb 25 22:51:45 mail postfix/postfix-script[4688]: fatal: the Postfix mail system is already running
2023-02-25 22:51:46,223 INFO exited: postfix (exit status 1; not expected)
2023-02-25 22:51:47,228 INFO spawned: 'postfix' with pid 4689
Feb 25 22:51:47 mail postfix/postfix-script[4695]: fatal: the Postfix mail system is already running
2023-02-25 22:51:47,256 INFO success: postfix entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
2023-02-25 22:51:48,257 INFO exited: postfix (exit status 1; not expected)
2023-02-25 22:51:49,262 INFO spawned: 'postfix' with pid 4696
Feb 25 22:51:49 mail postfix/postfix-script[4702]: fatal: the Postfix mail system is already running
2023-02-25 22:51:49,287 INFO success: postfix entered RUNNING state, process has stayed up for > than 0 seconds (startsecs)
2023-02-25 22:51:50,288 INFO exited: postfix (exit status 1; not expected)
```

### What did you expect to happen?

Clean restart of postfix.

### How do we replicate the issue?

1. Run `supervisorctl restart postfix` inside the container
2.
3.
...

### DMS version

edge

### What operating system is DMS running on?

Linux

### Which operating system version?

Debian 11

### What instruction set architecture is DMS running on?

AMD64 / x86_64

### What container orchestration tool are you using?

Docker

### Other relevant information

This is a regression introduced in https://github.com/docker-mailserver/docker-mailserver/pull/3033

---

Workaround:

```
supervisorctl stop postfix
service postfix stop
supervisorctl start postfix
```

### What level of experience do you have with Docker and mail servers?

- [ ] I am inexperienced with docker
- [X] I am rather experienced with docker
- [ ] I am inexperienced with mail servers
- [X] I am rather experienced with mail servers
- [ ] I am uncomfortable with the CLI
- [X] I am rather comfortable with the CLI

### Code of conduct

- [X] I have read this project's [Code of Conduct](https://github.com/docker-mailserver/docker-mailserver/blob/master/CODE_OF_CONDUCT.md) and I agree
- [X] I have read the [README](https://github.com/docker-mailserver/docker-mailserver/blob/master/README.md) and the [documentation](https://docker-mailserver.github.io/docker-mailserver/edge/) and I searched the [issue tracker](https://github.com/docker-mailserver/docker-mailserver/issues?q=is%3Aissue) but could not find a solution

**Comment by georglauterbach** [comment]
I think I noticed this as well and was able to recreate it!

**Comment by polarathene** [comment]
Probably not relevant but I am aware of this:

<details>
<summary>`/etc/init.d/postfix`:</summary>

```bash
#!/bin/sh -e

# Start or stop Postfix
#
# LaMont Jones <lamont@debian.org>
# based on sendmail's init.d script

### BEGIN INIT INFO
# Provides:          postfix mail-transport-agent
# Required-Start:    $local_fs $remote_fs $syslog $named $network $time
# Required-Stop:     $local_fs $remote_fs $syslog $named $network
# Should-Start:      postgresql mysql clamav-daemon postgrey spamassassin saslauthd dovecot
# Should-Stop:       postgresql mysql clamav-daemon postgrey spamassassin saslauthd dovecot
# Default-Start:     2 3 4 5
# Default-Stop:      0 1 6
# Short-Description: Postfix Mail Transport Agent
# Description:       postfix is a Mail Transport agent
### END INIT INFO

PATH=/bin:/usr/bin:/sbin:/usr/sbin
DAEMON=/usr/sbin/postfix
NAME=Postfix
TZ=
unset TZ

test -x $DAEMON && test -f /etc/postfix/main.cf || exit 0

. /lib/lsb/init-functions
#DISTRO=$(lsb_release -is 2>/dev/null || echo Debian)

enabled_instances() {
        postmulti -l -a | awk '($3=="y") { print $1}'
}

running() {
    INSTANCE="$1"
    if [ "X$INSTANCE" = X ]; then
            POSTMULTI=""
    else
            POSTMULTI="postmulti -i $INSTANCE -x "
    fi
    POSTCONF="${POSTMULTI} postconf"

    daemon_directory=$($POSTCONF -hx daemon_directory 2>/dev/null || echo /usr/lib/postfix/sbin)
    if ! ${POSTMULTI} $daemon_directory/master -t 2>/dev/null ; then
            echo y
    fi
}

case "$1" in
    start)
        log_daemon_msg "Starting Postfix Mail Transport Agent" postfix
        RET=0
        # for all instances that are not already running, handle chroot setup if needed, and start
        for INSTANCE in $(enabled_instances); do
            RUNNING=$(running $INSTANCE)
            if [ "X$RUNNING" = X ]; then
                /usr/lib/postfix/configure-instance.sh $INSTANCE
                CMD="/usr/sbin/postmulti -- -i $INSTANCE -x ${DAEMON}"
                if ! start-stop-daemon --start --exec $CMD quiet-quick-start; then
                    RET=1
                fi
            fi
        done
        log_end_msg $RET
    ;;

    stop)
        log_daemon_msg "Stopping Postfix Mail Transport Agent" postfix
        RET=0
        # for all instances that are not already running, handle chroot setup if needed, and start
        for INSTANCE in $(enabled_instances); do
            RUNNING=$(running $INSTANCE)
            if [ "X$RUNNING" != X ]; then
                CMD="/usr/sbin/postmulti -i $INSTANCE -x ${DAEMON}"
                if ! ${CMD} quiet-stop; then
                    RET=1
                fi
            fi
        done
        log_end_msg $RET
    ;;

    restart)
        $0 stop
        $0 start
    ;;

    force-reload|reload)
        log_action_begin_msg "Reloading Postfix configuration"
        if ${DAEMON} quiet-reload; then
            log_action_end_msg 0
        else
            log_action_end_msg 1
        fi
    ;;

    status)
        ALL=1
        ANY=0
        # for all instances that are not already running, handle chroot setup if needed, and start
        for INSTANCE in $(enabled_instances); do
            RUNNING=$(running $INSTANCE)
            if [ "X$RUNNING" != X ]; then
                ANY=1
            else
                ALL=0
            fi
        done
        # handle the case when postmulti returns *no* configured instances
        if [ $ANY = 0 ]; then
           ALL=0
        fi
        if [ $ALL = 1 ]; then
           log_success_msg "postfix is running"
           exit 0
        elif [ $ANY = 1 ]; then
           log_success_msg "some postfix instances are running"
           exit 0
        else
           log_success_msg "postfix is not running"
           exit 3
        fi
    ;;

    flush|check|abort)
        ${DAEMON} $1
    ;;

    *)
        log_action_msg "Usage: /etc/init.d/postfix {start|stop|restart|reload|flush|check|abort|force-reload|status}"
        exit 1
    ;;
esac

exit 0
```

</details>

`supervisorctl stop postfix` does not actually stop the process, just the supervisor unit, which would prevent the restart failure spam described in this issue. So technically I think supervisor needs to configure a proper `stop` action to `postfix stop`, and also for `restart` action to `postfix reload` or for a proper restart `postfix stop; postfix start-fg`.

**Comment by casperklein** [comment]
AFAIK all supervisord does is to send `SIGTERM` to the `postfix start-fg` process.

The wrapper script handled this with `traps`. After receiving SIGTERM, `service postfix stop` was executed.

---

I did some testing in a fresh debian container:

<details>
<summary>Details</summary>

```
# install postfix
apt install postfix

# check if postfix is running
ps aux | grep postfix

# not running - expected

# start postfix
postfix start-fg &

# check if postfix is running
ps aux | grep postfix

root         918  0.0  0.0   2484   516 pts/0    S    23:45   0:00 /bin/sh /usr/lib/postfix/sbin/postfix-script start-fg
root        1008  0.0  0.0  40056  6568 ?        Ss   23:45   0:00 /usr/lib/postfix/sbin/master -s
postfix     1009  0.0  0.0  40072  6632 ?        S    23:45   0:00 pickup -l -t unix -u -c
postfix     1010  0.0  0.0  40120  6556 ?        S    23:45   0:00 qmgr -l -t unix -u

# postfix is running - expected

# send SIGTERM to `postfix start-fg` process
kill %1

# check if postfix is running
ps aux | grep postfix
root        1008  0.0  0.0  40056  6568 ?        Ss   23:45   0:00 /usr/lib/postfix/sbin/master -s
postfix     1009  0.0  0.0  40072  6632 ?        S    23:45   0:00 pickup -l -t unix -u -c
postfix     1010  0.0  0.0  40120  6556 ?        S    23:45   0:00 qmgr -l -t unix -u

# some parts of postfix are still running

# stop remaining postfix parts
service postfix stop
Stopping Postfix Mail Transport Agent: postfix.

# check if postfix is running
ps aux | grep postfix

# not running - expected
```
</details>

---

For now, the only fix I can imagine is to reintroduce the wrapper script 😞

**Comment by polarathene** [comment]
http://supervisord.org/configuration.html#program-x-section-values

I am exploring some of the options with `stopsignal` and perhaps `pidproxy` (if there is a postfix pid file somewhere, or one that can be configured?). Those might be viable.

I found killing Postfix by PID manually was successful for me, but I typed the PID, instead of `%1` which I'm not familiar with :sweat_smile: I also didn't background(?) the `postfix start-fg` with `&`, but used separate TTY. I know that `postfix start-fg` doesn't seem to play nice with `CTRL+C` though, which is what supervisor with the `TERM` signal expects to work by default right?

Sadly supervisor doesn't seem to support different commands for `restart` / `stop`, only the signal handling to stop, and re-run the start command. It was requested long ago, but [intentionally decided against by the maintainers](https://github.com/Supervisor/supervisor/issues/147#issuecomment-52440691), and others [suggesting workaround to wrap in a bash script to trap signals](https://github.com/Supervisor/supervisor/issues/147#issuecomment-245688516) like you mentioned.

**Comment by georglauterbach** [comment]
This init system though..

---

Not 100% sure what to do about this. I've managed to get rid of the last wrapper today (upcoming PR); re-introducing one for Postfix would be sad IMO.

**Comment by polarathene** [comment]
Resolved, if the Postfix `command` value is updated to user pidproxy feature from `supervisorctl`, it'll work:

```
command=/usr/bin/pidproxy /var/spool/postfix/pid/master.pid /usr/sbin/postfix start-fg
```

It doesn't seem to need `/usr/bin` prefixing `pidproxy`, but it would not work properly without an explicit absolute path for `postfix` command.

**Comment by casperklein** [comment]
I can confirm this works.

**Comment by casperklein** [comment]
BTW I noticed this while restarting postfix:

```
Feb 26 01:20:02 mail postfix/postfix-script[5822]: warning: /var/spool/postfix/etc/hosts and /etc/hosts differ
Feb 26 01:20:02 mail postfix/postfix-script[5824]: warning: /var/spool/postfix/etc/ssl/certs/ssl-cert-snakeoil.pem and /etc/ssl/certs/ssl-cert-snakeoil.pem differ
```

```
diff /var/spool/postfix/etc/hosts /etc/hosts
7c7
< 192.168.0.2   mail.example.com mail
---
> 172.30.0.2    mail.example.com mail
```

```
ls -Ahl /var/spool/postfix/etc/hosts /etc/hosts
-rw-r--r-- 1 root root 180 Feb 26 01:06 /etc/hosts
-rw-r--r-- 1 root root 181 Feb 10 00:40 /var/spool/postfix/etc/hosts
```

```
ls -Ahl /var/spool/postfix/etc/
total 40K
-rw-r--r-- 1 root root    9 Feb 10 00:40 host.conf
-rw-r--r-- 1 root root  181 Feb 10 00:40 hosts
-rw-r--r-- 1 root root 2.3K Feb 10 00:40 localtime
-rw-r--r-- 1 root root  494 Feb 10 00:40 nsswitch.conf
-rw-r--r-- 1 root root   62 Feb 10 00:40 resolv.conf
-rw-r--r-- 1 root root  13K Feb 10 00:40 services
drwxr-xr-x 3 root root 4.0K Feb 10 00:40 ssl
```

None of these files seems to be touched anymore after Feb 10. Any idea what recent change could cause that?

**Comment by polarathene** [comment]
> None of these files seems to be touched anymore after Feb 10. Any idea what recent change could cause that?
> 
> Edit: Maybe the CHROOT change in `master.cf`?

This is something I have raised as a related concern a few times in a few issues / PRs. I haven't had time to investigate (_and keep forgetting to when I do_ :sweat_smile: ). I'm not sure that the `/var/spool/postfix` location is populated with the same content in v12 as past releases due to the removal of `chroot`.

Chances are a folder or perhaps a few in `/var/spool/postfix` would no longer be generated, or would be different content.

With the chroot at least, I recall that Postfix would copy content from various locations, such as the `/etc/hosts` file, as it needed it's own copies in it's chroot. This would also only be sync'd when Postfix was started, not when performing a `reload` like we've introduced in v12 for change detection.

Thus as you've noticed, this may be a slight hiccup with upgrades from earlier releases when you've got `/var/mail-state` mount present. TBH, I don't know how relevant that is to persist, whomever decided to persist the location perhaps did so because there was something worthwhile persisting in `/var/spool/postfix` :man_shrugging: (_There is some top-level dirs for state I think like `maildrop/` and `incoming/`, but this is interleaved with folders for the chroot environment (`dev/`, `etc/`, `lib/`, `usr/`) and run-time like `pid/`_)

**Comment by polarathene** [comment]
Here is the content from a fresh image of my last master branch build (earlier this month), no `/var/mail-state` mount via `docker run --rm -it --hostname example.test --env SMTP_ONLY=1 <image>`:

**`ls -lah /var/spool/postfix/*`:**

<details>
<summary>v12 (no volume mount)</summary>

```
/var/spool/postfix/active:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/bounce:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/corrupt:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/defer:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/deferred:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/dev:
total 0
drwxr-xr-x 1 root root 17 Feb 26 00:43 .
drwxr-xr-x 1 root root 82 Feb 26 00:43 ..
srw-rw-rw- 1 root root  0 Feb 26 00:43 log

/var/spool/postfix/etc:
total 0
drwxr-xr-x 2 root root  6 Oct 12 02:30 .
drwxr-xr-x 1 root root 82 Feb 26 00:43 ..

/var/spool/postfix/flush:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/hold:
total 0
drwx------ 2 postfix root  6 Feb 26 00:43 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/incoming:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/lib:
total 0
drwxr-xr-x 2 root root  6 Oct 12 02:30 .
drwxr-xr-x 1 root root 82 Feb 26 00:43 ..

/var/spool/postfix/maildrop:
total 0
drwx-wx--T 2 postfix postdrop  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root     82 Feb 26 00:43 ..

/var/spool/postfix/pid:
total 4.0K
drwxr-xr-x 1 root root 24 Feb 26 00:43 .
drwxr-xr-x 1 root root 82 Feb 26 00:43 ..
-rw------- 1 root root 33 Feb 26 00:43 master.pid

/var/spool/postfix/private:
total 4.0K
drwx------ 1 postfix root    4.0K Feb 26 00:43 .
drwxr-xr-x 1 root    root      82 Feb 26 00:43 ..
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 anvil
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 bounce
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 defer
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 discard
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 dnsblog
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 error
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 lmtp
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 local
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 policyd-spf
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 proxymap
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 proxywrite
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 relay
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 retry
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 rewrite
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 scache
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 smtp
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 smtp-amavis
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 smtpd
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 tlsmgr
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 tlsproxy
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 trace
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 verify
srw-rw-rw- 1 postfix postfix    0 Feb 26 00:43 virtual

/var/spool/postfix/public:
total 0
drwx--s--- 1 postfix postdrop 95 Feb 26 00:43 .
drwxr-xr-x 1 root    root     82 Feb 26 00:43 ..
srw-rw-rw- 1 postfix postdrop  0 Feb 26 00:43 cleanup
srw-rw-rw- 1 postfix postdrop  0 Feb 26 00:43 flush
prw--w--w- 1 postfix postdrop  0 Feb 26 00:43 pickup
srw-rw-rw- 1 postfix postdrop  0 Feb 26 00:43 qmgr
srw-rw-rw- 1 postfix postdrop  0 Feb 26 00:43 sender-cleanup
srw-rw-rw- 1 postfix postdrop  0 Feb 26 00:43 showq

/var/spool/postfix/saved:
total 0
drwx------ 2 postfix root  6 Feb 14 23:11 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/trace:
total 0
drwx------ 2 postfix root  6 Feb 26 00:43 .
drwxr-xr-x 1 root    root 82 Feb 26 00:43 ..

/var/spool/postfix/usr:
total 0
drwxr-xr-x 3 root root 17 Feb 14 23:11 .
drwxr-xr-x 1 root root 82 Feb 26 00:43 ..
drwxr-xr-x 4 root root 35 Feb 14 23:11 lib
```

</details>

<details>
<summary>v11.3 (no volume mount)</summary>

```
/var/spool/postfix/active:
total 0
drwx------ 1 postfix root   6 Feb 26 01:01 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/bounce:
total 0
drwx------ 2 postfix root   6 Dec 21 10:26 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/corrupt:
total 0
drwx------ 2 postfix root   6 Dec 21 10:26 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/defer:
total 0
drwx------ 2 postfix root   6 Dec 21 10:26 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/deferred:
total 0
drwx------ 2 postfix root   6 Dec 21 10:26 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/dev:
total 0
drwxr-xr-x 1 root root   46 Feb 26 01:01 .
drwxr-xr-x 1 root root  161 Feb 26 01:01 ..
srw-rw-rw- 1 root root    0 Feb 26 01:01 log
crw-rw-rw- 1 root root 1, 8 Feb 26 01:01 random
crw-rw-rw- 1 root root 1, 9 Feb 26 01:01 urandom

/var/spool/postfix/etc:
total 36K
drwxr-xr-x 1 root root 120 Feb 26 01:01 .
drwxr-xr-x 1 root root 161 Feb 26 01:01 ..
-rw-r--r-- 1 root root   9 Feb 26 01:01 host.conf
-rw-r--r-- 1 root root 182 Feb 26 01:01 hosts
-rw-r--r-- 1 root root 118 Feb 26 01:01 localtime
-rw-r--r-- 1 root root 494 Feb 26 01:01 nsswitch.conf
-rw-r--r-- 1 root root  49 Feb 26 01:01 resolv.conf
-rw-r--r-- 1 root root 13K Feb 26 01:01 services
drwxr-xr-x 3 root root  19 Feb 26 01:01 ssl

/var/spool/postfix/flush:
total 0
drwx------ 2 postfix root   6 Dec 21 10:26 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/hold:
total 0
drwx------ 2 postfix root   6 Feb 26 01:01 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/incoming:
total 0
drwx------ 1 postfix root   6 Feb 26 01:01 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/lib:
total 4.0K
drwxr-xr-x 1 root root   30 Feb 26 01:01 .
drwxr-xr-x 1 root root  161 Feb 26 01:01 ..
drwxr-xr-x 2 root root 4.0K Feb 26 01:01 x86_64-linux-gnu

/var/spool/postfix/maildrop:
total 0
drwx-wx--T 1 postfix postdrop   6 Feb 26 01:01 .
drwxr-xr-x 1 root    root     161 Feb 26 01:01 ..

/var/spool/postfix/pid:
total 4.0K
drwxr-xr-x 1 root root  62 Feb 26 01:01 .
drwxr-xr-x 1 root root 161 Feb 26 01:01 ..
-rw------- 1 root root  33 Feb 26 01:01 master.pid
-rw------- 1 root root   0 Feb 26 01:01 unix.cleanup
-rw------- 1 root root   0 Feb 26 01:01 unix.local

/var/spool/postfix/private:
total 4.0K
drwx------ 1 postfix root    4.0K Feb 26 01:01 .
drwxr-xr-x 1 root    root     161 Feb 26 01:01 ..
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 anvil
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 bounce
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 defer
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 discard
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 dnsblog
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 error
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 lmtp
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 local
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 policyd-spf
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 proxymap
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 proxywrite
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 relay
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 retry
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 rewrite
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 scache
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 smtp
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 smtp-amavis
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 smtpd
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 tlsmgr
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 tlsproxy
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 trace
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 verify
srw-rw-rw- 1 postfix postfix    0 Feb 26 01:01 virtual

/var/spool/postfix/public:
total 0
drwx--s--- 1 postfix postdrop  95 Feb 26 01:01 .
drwxr-xr-x 1 root    root     161 Feb 26 01:01 ..
srw-rw-rw- 1 postfix postdrop   0 Feb 26 01:01 cleanup
srw-rw-rw- 1 postfix postdrop   0 Feb 26 01:01 flush
prw--w--w- 1 postfix postdrop   0 Feb 26 01:01 pickup
srw-rw-rw- 1 postfix postdrop   0 Feb 26 01:01 qmgr
srw-rw-rw- 1 postfix postdrop   0 Feb 26 01:01 sender-cleanup
srw-rw-rw- 1 postfix postdrop   0 Feb 26 01:01 showq

/var/spool/postfix/saved:
total 0
drwx------ 2 postfix root   6 Dec 21 10:26 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/trace:
total 0
drwx------ 2 postfix root   6 Feb 26 01:01 .
drwxr-xr-x 1 root    root 161 Feb 26 01:01 ..

/var/spool/postfix/usr:
total 0
drwxr-xr-x 1 root root  17 Dec 21 10:26 .
drwxr-xr-x 1 root root 161 Feb 26 01:01 ..
drwxr-xr-x 1 root root  22 Dec 21 10:26 lib
```

</details>

Difference, v11.3 has the additional content:
- `dev/` has `random` + `urandom`.
- `etc/` has same content you've listed above.
- `lib/` has `x86_64-linux-gnu/` with files for gcc, nss, resolv.
- `pid/` additionally has `unix.cleanup` and `unix.local`, both zero sized.

Presumably none of that is relevant going forward with v12. Interesting that you're getting those warnings though :thinking: I wouldn't have thought it'd care about those copies if it's not meant to use them without chroot. I haven't investigated if it'd cause any issues, perhaps release notes should mention that it shouldn't be relevant, or we could look if `main.cf` has a setting that could be adjusted (perhaps [`queue_directory`](https://www.postfix.org/postconf.5.html#queue_directory)?).

**Comment by georglauterbach** [comment]
I actually already removed them a while ago as I kept getting the same "warnings". The directories are being created again afterwards, but as @polarathene has shown, mostly without content.

**Comment by casperklein** [comment]
Thanks for your detailed explanation @polarathene (as always) 👍

Maybe we should add a cleanup routine which can be removed after some release cycles: `rm -rf /var/spool/postfix/{dev,etc,lib,usr}`

**Comment by georglauterbach** [comment]
> Maybe we should add a cleanup routine which can be removed after some release cycles: `rm -rf /var/spool/postfix/{dev,etc,lib,usr}`

I will take care of that.

**Comment by casperklein** [comment]
I've removed the four directories and the warnings are gone 👍
